### PR TITLE
Make `TPESampler` pruning aware.

### PR DESCRIPTION
**Pull request by sile**
This PR makes `TPESampler` pruning aware.
More specifically, this PR changes `TPESampler` to account for the pruned steps and intermediate values of pruned trials in addition to complete ones.

The following benchmark results show that this modification can dramatically improve optimization performance when `TPESampler` is used with pruning.

## (1) Benchmark Results

Please refer to [Tabular Benchmarks for Joint Architecture and Hyperparameter Optimization](https://arxiv.org/abs/1905.04970) for the datasets and search space used by this benchmarking.

### (1-1) Pruning Benchmark

result0 result1

result2result3

#### Ranking

`$ cat *.json | grep -v noprune | kurobako stats ranking`

| Sampler                                       | Borda | Firsts |
|:---------------------------------------------|------:|-------:|
| (a) optuna#master:minimize:asha              |     4 |      0 |
| (b) optuna#master:minimize:median            |     0 |      0 |
| (c) optuna#pruning-aware-tpe:minimize:asha   |     9 |      4 |
| (d) optuna#pruning-aware-tpe:minimize:median |     8 |      3 |

#### Ranking

noprune-0noprune-1

noprune-2 noprune-3

`$ cat *.json | grep noprune | kurobako stats ranking`

| Samplers                                       | Borda | Firsts |
|:----------------------------------------------|------:|-------:|
| (a) optuna#master:maximize:noprune            |     0 |      4 |
| (b) optuna#master:minimize:noprune            |     0 |      3 |
| (c) optuna#pruning-aware-tpe:maximize:noprune |     1 |      4 |
| (d) optuna#pruning-aware-tpe:minimize:noprune |     0 |      4 |

In no pruning cases, there are no statistically significant performance degradations by this modification.

## (2) Benchmark Commands

```console
// Install packages
$ cargo install kurobako
$ kurobako --version
kurobako 0.0.14

$ pip install git+http://github.com/sile/kurobako-py.git
$ pip install git+http://github.com/pfnet/optuna.git@pruning-aware-tpe2

// Download datasets
$ wget http://ml4aad.org/wp-content/uploads/2019/01/fcnet_tabular_benchmarks.tar.gz
$ tar xf fcnet_tabular_benchmarks.tar.gz

// Run benchmarks
$ kurobako benchmark --problems $(kurobako problem-suite fc-net fcnet_tabular_benchmarks) --solvers $(kurobako solver --tag pruning-aware-tpe:minimize:asha optuna --pruner asha --asha-min-resource 3 --asha-reduction-factor 3) --budget 50 --iterations 40 --evaluation-points 3 9 27 81 100 | kurobako run > pruning-aware-tpe-minimize-asha.json
$ kurobako benchmark --problems $(kurobako problem-suite fc-net fcnet_tabular_benchmarks) --solvers $(kurobako solver --tag master:minimize:median optuna) --budget 50 --iterations 40 --evaluation-points 10 20 30 40 50 60 70 80 90 100 | kurobako run > master-minimize-median.json
$ kurobako benchmark --problems $(kurobako problem-suite fc-net fcnet_tabular_benchmarks) --solvers $(kurobako solver --tag pruning-aware-tpe:maximize:noprune optuna --maximize) --budget 50 --iterations 40 --evaluation-points 100 | kurobako run > pruning-aware-tpe-maximize-noprune.json
$ kurobako benchmark --problems $(kurobako problem-suite fc-net fcnet_tabular_benchmarks) --solvers $(kurobako solver --tag pruning-aware-tpe:maximize:noprune optuna) --budget 50 --iterations 40 --evaluation-points 100 | kurobako run > pruning-aware-tpe-minimize-noprune.json

// Plot results
$ cat *.json | grep -v noprune | kurobako select --problems fcnet_slice_localization_data | kurobako plot -o /tmp/plot-result/ --ymin 0.0 --ymax 0.002
$ cat *.json | grep -v noprune | kurobako select --problems fcnet_protein_structure_data | kurobako plot -o /tmp/plot-result/ --ymin 0.2 --ymax 0.4
$ cat *.json | grep -v noprune | kurobako select --problems fcnet_parkinsons_telemonitoring_data | kurobako plot -o /tmp/plot-result/ --ymin 0 --ymax 0.05
$ cat *.json | grep -v noprune | kurobako select --problems fcnet_naval_propulsion_data | kurobako plot -o /tmp/plot-result/ --ymin 0 --ymax 0.002
```

### CODE REVIEW

File: optuna/storages/base.py [hunk 1]
```diff
@@ -200,20 +200,6 @@ def get_trial_system_attrs(self, trial_id):
 
         return self.get_trial(trial_id).system_attrs
 
-    # Methods for the TPE sampler
-
-    def get_trial_param_result_pairs(self, study_id, param_name):
-        # type: (int, str) -> List[Tuple[float, float]]
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/storages/base.py]: Thanks! Removed: https://github.com/pfnet/optuna/pull/439/commits/abbb8286df27cc30098f795be980978d78006a08

**Comment by sile** [comment]
@toshihikoyanase 

> I think it is helpful for the future development of `TPESampler` if you add an explanation of `_get_observation_pair`.

Thank you for your suggestion. I added documentation at the commit https://github.com/pfnet/optuna/pull/439/commits/3ba0d19f37f021d49346a5a79435580bc0390d2d.

File: optuna/samplers/tpe/sampler.py [hunk 2]
```diff
@@ -448,6 +448,20 @@ def _log_normal_cdf(cls, x, mu, sigma):
 
 def _get_observation_pairs(study, param_name):
     # type: (InTrialStudy, str) -> List[Tuple[float, Tuple[float, float]]]
+    """Get observation pairs from the study.
+
+       This function collects observation pairs from the trials of the study.
+       The trials that doesn't contain the parameter named ``param_name`` are excluded
```
Comment (toshihikoyanase) [inline optuna/samplers/tpe/sampler.py]: Thank you for adding the comments. I understood the behavior and the usage of `_get_observation_pairs` clearly.

I found a small mistake here.
`The trials that doesn't contain` seems to be a typo for `The trials that don't contain`.

Comment (sile) [reply optuna/samplers/tpe/sampler.py]: I already fixed the mistake https://github.com/pfnet/optuna/pull/439/commits/8f697a5b8579f99a5987273a4a341eafe56f9dc9.
Sorry for my carelessness.

**Pull request merged**